Sources/NetworkProtection/Settings/Extensions/UserDefaults+dnsSettings.swift

```diff
@@ -0,0 +1,80 @@
+//
+//  UserDefaults+dnsSettings.swift
+//
+//  Copyright © 2024 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Combine
+import Foundation
+
+extension UserDefaults {
+    final class StorableDNSSettings: NSObject, Codable {
+        let usesCustomDNS: Bool
+        let dnsServers: [String]
+
+        init(usesCustomDNS: Bool = false, dnsServers: [String] = []) {
+            self.usesCustomDNS = usesCustomDNS
+            self.dnsServers = dnsServers
+        }
+    }
+
+    private var dnsSettingKey: String {
+        "dnsSettingStorageValue"
+    }
+
+    private static func dnsSettingsFromStorageValue(_ value: StorableDNSSettings) -> NetworkProtectionDNSSettings {
+        guard value.usesCustomDNS else { return .default }
+        return .custom(value.dnsServers)
```

Comment:
Actually, related to the previous comment, here is where we should likely have some recovery logic in the case that `value.usesCustomDNS` is true but `value.dnsServers` is empty for some reason. (Probably it can have an assertion failure as well, to let us know locally if this ever happens.)